zerver/actions/message_send.py

```diff
@@ -864,7 +865,27 @@ def do_send_messages(
             send_request.message, send_request.rendering_result.potential_attachment_path_ids
         ):
             send_request.message.has_attachment = True
-            send_request.message.save(update_fields=["has_attachment"])
+            update_fields = ["has_attachment"]
+
+            # Lock the ImageAttachment rows that we pulled when rendering
+            # outside of the transaction; they may be out of date now, and we
+            # need to post-process the rendered_content to swap out the
+            # spinner.  We would ideally take a `FOR SHARE` lock here
+            # (https://www.postgresql.org/docs/current/explicit-locking.html#LOCKING-ROWS),
+            # which would block UPDATEs but not other `FOR SHARE`, but Django
+            # does not support this yet: (https://code.djangoproject.com/ticket/10088)
+            assert send_request.message.rendered_content is not None
+            previews = get_user_upload_previews(
+                send_request.message.realm_id, send_request.message.content, lock=True
+            )
+            new_rendered_content = rewrite_thumbnailed_images(
+                send_request.message.rendered_content, previews
+            )
+            if new_rendered_content is not None:
+                send_request.message.rendered_content = new_rendered_content
+                update_fields.append("rendered_content")
+
+            send_request.message.save(update_fields=update_fields)
```

Comment:
I wonder if there's some optimization we can do, wherein we keep track of whether the rendered content in the message has any loading spinners waiting for thumbnailing, and only do this extra work if so?

Reply:
Yeah, I think we can track that bit of information.